### PR TITLE
Fix race where it may cause a hang

### PR DESCRIPTION
**Pull request by noobpwnftw**
In case of the notifier thread doesn't find a waiter, and a subsequent wait consumes the notify in the same thread, it will give one back, if the predicate success, no more notify should fire, if it fails, this will essentially turn the offending thread into a spin loop waiting on the other threads to enter wait.

This patch also disallow default copy operators for Mutex(as per C++11 implementations), although that doesn't really affect us.
Using compiler default mutexes had issues described in the original comment, as tested recently they still cause minor slowdowns on MINGW compiles.

No functional change

### CODE REVIEW

**Comment by vondele** [comment]
Nice catch. 

However, as an alternative, can we investigate if we should just get rid of the custom Mutex implementation for windows ? The code is from 2015, and I'm guessing Mutex is much less performance critical now than it was before lazy smp ....

**Comment by noobpwnftw** [comment]
@vondele I did the hang test with the compiler default mutexes during investigation, the results are as expected(no problems with old MINGW and other compilers). However I didn't ask a test for performance difference. I guess the MINGW guys fall for the same problem we have: wrapped the native objects up without noticing that those are in fact non-copyable.
Before they fix that, compiling on latest MINGW is largely broken regardless of what we do.

**Comment by mooskagh** [comment]
But where was this Mutex actually copied?
I cannot find any relevant place in the code. Also in most of the cases I'd expect deleting copy constructor/assignment operator would cause build to break, I can only think of exotic cases where it could actually change behavior instead.

Am I missing something? For me it looks like this change shouldn't change anything at all (if it still compiles).

**Comment by noobpwnftw** [comment]
@mooskagh In condition variables the wait function takes the mutex as a reference, my guess that's where it broke. There are also a few internal copies of mutexes using `std::move` within the implementation.
Evidence: https://github.com/official-stockfish/Stockfish/issues/2291#issuecomment-531596621
That was built to use our own Mutex(without removing the copy operators) and a conditional variable implemented as per standard libraries using semaphores, which broke both compiles.
Results get blurred when the conditional variable implementation in old `libwinpthread` has its own semaphores, thus probably will never trigger a race condition where you are not supposed to copy `CRITICAL_SECTION`s (https://docs.microsoft.com/zh-cn/windows/win32/api/synchapi/nf-synchapi-initializecriticalsection, search for 'A critical section object cannot be moved or copied.')

**Comment by vondele** [comment]
@noobpwnftw if performance is OK, we should just remove the custom code.

**Comment by mooskagh** [comment]
Could you point me to an example in the code when `Mutex` object would be copied before the fix?

**Comment by noobpwnftw** [comment]
@vondele I agree.
@mooskagh We don't, and that's why there is a problem.

**Comment by noobpwnftw** [comment]
So there is one more suspect:
In case of a wake up that did not satisfy the predicate function, wake up again to avoid a data race.

**Comment by noobpwnftw** [comment]
By reading code I think it was intended to do something like "wait_once" for each thread, but in fact it just so happens that there may be the some threads took the notify more than once(while the check of predicate immediately following the wait didn't satisfy, so it will consume more than one notify). My ugly way of doing things essentially forces one notify before each check of predicate, so that they can never run out of signals unless they leave the wait function.
The signal leftovers should be harmless, because if there is no waiting thread, `notify_one` will do nothing.

Reference:
https://stackoverflow.com/questions/42714632/notify-one-in-c-thread-waking-up-more-than-one-thread

**Comment by snicolet** [comment]
What happens if we make the searching flag atomic, btw?

**Comment by noobpwnftw** [comment]
@snicolet Probably that would make it slower.

**Comment by snicolet** [comment]
I would much prefer to use an atomic if there is a data race, rather than "fixing" it by double notify!
Btw, can you write a detailled example of scenario leading to the race? With two threads should be enough :-)

**Comment by noobpwnftw** [comment]
I understand, but I can see there are good reasons why we don't, and bugs are extremely rare.

For such a race scenario, consider:
1. All search threads are in waiting state for predicate `searching = true` and a signal.
The implementation of a conditional variable would look like this:

```
wait(external_lock, predicate_fn)
{
    while(!predicate_fn)
    {
        external_lock.unlock();
        wait_for_signal();               <--- we are here
        external_lock.lock();
    }
}
```
2. UCI thread issues `notify_one` once for each thread.
3. One thread receives one signal, but in its view, `searching` is not updated, predicate check fails and wait for another signal.
4. Another signal is never received because we only issue once per thread/wait.

**Comment by Sopel97** [comment]
atomic bools that are not loaded/stored simultanously by many threads have no noticable overhead on modern hardware

**Comment by vondele** [comment]
The performance of accessing the 'searching' variable is not performance critical, so could be atomic if needed. 

However, I'm not so sure we have a race there, every access to the variable searching seems to be protected by a mutex. Isn't that enough to guarantee consistency? 

Also, I would expect our thread-sanitizer CI to notice this if we have a race there.

**Comment by Sopel97** [comment]
"every access to the variable searching seems to be protected by a mutex. Isn't that enough to guarantee consistency?" it may not be enough with compiler optimisations. The compiler doesn't have to generate loads if it already has the value.

**Comment by vondele** [comment]
@Sopel97 I'm not sure your answer is correct, I think they imply memory barriers that compilers need to respect, see also:
https://stackoverflow.com/questions/24137964/does-pthread-mutex-lock-contains-memory-fence-instruction

**Comment by noobpwnftw** [comment]
@vondele Since we have tested on using pure compiler mutexes and on custom mutexes, they both hang with the latest `libwinpthread`, my MSVC compile using a custom CV implementation based on Semaphores would still fail, which is free from any problems from the MINGW side.

**Comment by mooskagh** [comment]
Could someone who is able to reproduce that take stack traces of all threads while it hangs? (and also there's a proc dump from TCEC. If someone can reproduce exactly the same build, it's also possible to see thread state there).

I expect that once we see a stack trace, it should be pretty easy to find out what's happening.
Just trying to guess and changing arbitrary lines doesn't seem to be very productive.

**Comment by snicolet** [comment]
@mooskagh 
+1

**Comment by noobpwnftw** [comment]
During master PGO compile on MINGW64:
image
[stockfish.zip](https://github.com/official-stockfish/Stockfish/files/3616916/stockfish.zip)

**Comment by vondele** [comment]
I would propose to split the PR. As it now starts addressing two independent things

1) our Mutex implementation needs to be explicitly neither copyable nor movable, which we can do adding the two lines to the class, or by removing the full implementation. The latter option might be 2% slower, but removes quite some code.

2) Fix a potential race elsewhere, possibly the searching variable.

**Comment by noobpwnftw** [comment]
I can occasionally reproduce the hang and there is the full memory dump and the corresponding binary file you wanted.

**Comment by snicolet** [comment]
>I can occasionally reproduce the hang [...]

YEAH! I posted a branch with debug traces in this comment, and various parametrizations of the search behaviour, do you manage to make it hang on you system?

https://github.com/official-stockfish/Stockfish/issues/2291#issuecomment-531545670
https://github.com/snicolet/Stockfish/tree/search_explosion2

**Comment by noobpwnftw** [comment]
@snicolet No, can't produce a hang with your branch(while master hang is easy to produce).

**Comment by mooskagh** [comment]
> I can occasionally reproduce the hang and there is the full memory dump and the corresponding binary file you wanted.

Yeah, would be useful. I don't have windows but still can try opening the dump.

It can be either binary file with symbols compiled in (it's possible to do that even for release builds), or build+symbols file.
Also it would be useful to know on which exactly source snapshot it was compiled (so that if symbols data has line numbers, it's possible to match to the source code).

To get the dump of the process:
```bash
procdump.exe -ma <process id or process_filename.exe>  # to get full dump (may be large)
procdump.exe -mm <process id or process_filename.exe>  # to get dump without heap (probably enough)
```

**Comment by noobpwnftw** [comment]
@mooskagh https://github.com/official-stockfish/Stockfish/pull/2307#issuecomment-531824016
Binary file is with symbols, and it is a full process memory dump. It uses this branch without the additions of `notify_one`.

**Comment by mooskagh** [comment]
Just to confirm, is there somewhere dump attached? I cannot find it anywhere.

**Comment by vondele** [comment]
@mooskagh click on stockfish.zip in the comment above https://github.com/official-stockfish/Stockfish/pull/2307#issuecomment-531824016

**Comment by mooskagh** [comment]
Thanks, I didn't notice it.
I can open the dump, but it says that neither dump nor .exe file contains symbols.
So I would be useful to also have .pdb file from that build.

**Comment by noobpwnftw** [comment]
It's a MINGW compile, it is not possible to have PDB symbols, while it should have them in DWARF format embedded in the .exe file.

**Comment by mooskagh** [comment]
Indeed, I managed to extract .pdb with cv2pdb.exe, but VS complains that it doesn't match the binary..

**Comment by noobpwnftw** [comment]
I'm trying to reproduce one with MSVC compiler, so that it may be easier.
Problem is, master branch as-is, it is very hard to reproduce, it wasn't discovered before I was experimenting with non-default `std::mutex`(that critical section wrapper) and my own CV implementation.

**Comment by mooskagh** [comment]
I managed to open dump but not load symbols in windows.
Linux gdb can open .exe file and load symbols, but cannot load windows dump file.

Trying to manually match addesses and disassembly there.
Progress so far:
Main thread is hanging here:
https://github.com/official-stockfish/Stockfish/blob/master/src/thread.cpp#L96
Two other threads hanging here:
https://github.com/official-stockfish/Stockfish/blob/master/src/thread.cpp#L118

Probably nothing new so far, looking further.

**Comment by noobpwnftw** [comment]
image

[Hang_MSVC.zip](https://github.com/official-stockfish/Stockfish/files/3617669/Hang_MSVC.zip)

This is a hang(very rare) during bench, compiled with MSVC from https://github.com/noobpwnftw/Stockfish/tree/hang.

Binary, PDB, full memory dump included. I'm not some saint that writes bulletproof code, but those wrappers of Windows API are dumb enough not to blow up.

**Comment by mooskagh** [comment]
Thanks, that works perfectly!

Stack traces:

Main thread:
`searching=true` (waiting for false), `exit=false`
```
 	[External Code]	
>	Stockfish.exe!Thread::wait_for_search_finished() Line 96	C++
 	Stockfish.exe!`anonymous namespace'::bench(Position & pos={...}, std::basic_istream<char,std::char_traits<char> > & args, std::unique_ptr<std::deque<StateInfo,std::allocator<StateInfo> >,std::default_delete<std::deque<StateInfo,std::allocator<StateInfo> > > > & states={...}) Line 163	C++
 	Stockfish.exe!UCI::loop(int argc=2, char * * argv) Line 235	C++
 	Stockfish.exe!main(int argc=2, char * * argv=0x000001ddb4593260) Line 52	C++
 	[External Code]	
```

Thread 2:
`searching=false` (waiting for true) `exit=false`
```
 	[External Code]	
>	Stockfish.exe!Thread::idle_loop() Line 118	C++
 	Stockfish.exe!std::_LaunchPad<std::unique_ptr<std::tuple<void (__cdecl Thread::*)(void),Thread *>,std::default_delete<std::tuple<void (__cdecl Thread::*)(void),Thread *> > > >::_Go() Line 230	C++
 	Stockfish.exe!std::_Pad::_Call_func(void * _Data) Line 209	C++
 	Stockfish.exe!thread_start<unsigned int (__cdecl*)(void * __ptr64)>(void * const parameter=0x0000000000000000) Line 115	C++
 	[External Code]	
```

Thread 3:
`searching=true` (waiting for true) `exit=false`
```
 	[External Code]	
>	Stockfish.exe!Thread::idle_loop() Line 118	C++
 	Stockfish.exe!std::_LaunchPad<std::unique_ptr<std::tuple<void (__cdecl Thread::*)(void),Thread *>,std::default_delete<std::tuple<void (__cdecl Thread::*)(void),Thread *> > > >::_Go() Line 230	C++
 	Stockfish.exe!std::_Pad::_Call_func(void * _Data) Line 209	C++
 	Stockfish.exe!thread_start<unsigned int (__cdecl*)(void * __ptr64)>(void * const parameter=0x0000000000000000) Line 115	C++
 	[External Code]	
```

**Comment by mooskagh** [comment]
Main thread is waiting for the same Thread object as Thread 3.

So it seems to me what happens here is:
1. Thread 1 waits for (searching==false)
2. Thread 3 waits for (searching==true)
3. Something (I'm still not sure how SF synchronization works) sets searching=true, and calls notify_one().
4. Thread 1 wakes up, checks that searching is still false, and falls asleep.
5. Thread 3 doesn't wake up as it's notify_one() and not notify_all().

Do I miss something?

**Comment by noobpwnftw** [comment]
So the theory is that, provided my binary uses only native Windows API to implement mutexes and conditional variables and compiles with MSVC, which implies proper barriers around modifying/testing `searching` flag, either my wrapper is flawed, or it's a missing `notify_one`. I believe making `searching` atomic won't fix anything.

**Comment by noobpwnftw** [comment]
The missing `notify_one` can be there is no waiting thread, as per specification, it will not do anything if there are no waiting threads, however, I'm using Semaphores which will count the number of releases and alert the exact same number of waits, so do almost all implementations, so this possibility cannot exist.

**Comment by mooskagh** [comment]
How can non-main thread start searching, I'm missing in the code.
`searching=true` can only be set either initially in constructor, or by `start_searching()`, which is only called in destructor (useless for our case) and from ThreadPool::startThinking() **only** for main thread.

non-main `Thread::idle_loop()` sets it to false and starts to wait it to be `true`. Which code can possibly set it to true?

**Comment by noobpwnftw** [comment]
@mooskagh `searching` is a per-thread flag, set by calling all but `MainThread`'s `start_searching`.

**Comment by vondele** [comment]
at line 241 of search.cpp.

**Comment by snicolet** [comment]
I am away from computer at the moment and writing from a phone, but would it possible to replace all the ‘lock-guard lk(mutex)’ and ‘unique-lock lk(mutex)’ by explicit calls to mutex.lock() and mutex.unlock(), and see what happens?

Lock-guard and unique-lock have implicit release of the locks at the end of the life of their object, than make it sometimes tricky to reason sbout

**Comment by mooskagh** [comment]
`notify_one()` indeed can be lost if there are no waiting threads, but we signalling variable is always modified before notification, and wait() always checks predicate first before starting to wait.

May I ask why `notify_one()` and not `notify_all()`?

**Comment by mooskagh** [comment]
I'm sure that locks are not a problem here.

For me it sounds the most likely that the reason that there are two wait()s waiting for opposite predicates, and `notify_one()` call wakes up a wrong one. Adding more notify_one() won't help.

var=false
1. Thread 1 waits for var=true
2. Thread 2 waits for var=false
3. notify_one() happens to wake up thread 1, which checks the predicate, and continues to sleep. Even if you call it million times after every line in the code, it can still all go to thread 1.

So basically in a situation when two threads are waiting for opposite predicates and there are no `notify_all()` is automatically a deadlock (you may be lucky and wake up another thread, but you cannot rely on that).

**Comment by snicolet** [comment]
Btw, lock-guard lk(mutex) takes à copy of mutex, doesn’it, and then locks the copy? Would’t It be better to lock the original mutex?

**Comment by mooskagh** [comment]
No, lock parameter is reference, not value. There's no copying.

**Comment by vondele** [comment]
@mooskagh ... I can follow your argument, but I think the question is now how we would come in this state of waiting for opposite values of searching. I think the notify_one calls always exactly match the number of waiting threads, and they are all waiting in the same state... at least that would be the idea, but of course the implementation might be wrong.

**Comment by mooskagh** [comment]
I'm still trying to find a scenario how that can happen, but just to clarify, there's not much interoperation between threads, as every Thread object has it's own independent cv, mutex, searching and exit variables.

I could come up with hypothetical (with some assumptions) scenario how it could stuck with `searching=false`, but not `true` for now. For `false` it is:

1. searching = true
2. UCI thread calls `wait_for_search_finished()` which waits for searching=false
3. Worker thread in `idle_loop()` sets `searching=false` and does `notify_one()`. However as mutex is locked at that moment, UCI thread doesn't wake up yet.
4. Worker thread starts to wait for `searching=true` and unlocks mutex.
5. (assumption and speculation) Instead of UCI thread waking up to check, as this the same worker thread is waiting for the same variable, it uses it's own `notify_one()` to wake up and check that his predicate fails.
6. Everyone is asleep.

But we don't have hanging with `searching=false`, we have hanging with `searching=true`..

Also I must say it's the first time ever I see same condition variable waiting for two different predicates, I'm not sure they are even designed for that?..
I would use two different cv for that.. maybe?..

**Comment by noobpwnftw** [comment]
So how should one implement `notify_all` with non-exotic Windows API? I fear that the implementation inside `libwinpthread` is a ReleaseSemaphore with the # of waiters...

**Comment by vondele** [comment]
> @snicolet GOT ONE!
> 
> image
> 
> It reads "the last-1 thread does not seem to get a signal".
looking at this debug output, one can see that thread 6 didn't start the search (i.e. no 'setting maximum_depth X' output), so clearly the waiting on it to finish doesn't work. Do you happen to have more output for this (to see how that process finished the previous search)?

**Comment by noobpwnftw** [comment]
@vondele the last iteration was normal, all threads finish search before `bestmove` was sent. I ran with `concurrency = 1`, only one engine runs at a time.

**Comment by vondele** [comment]
in the output what we see if the bestmove of the other player. (i.e. hang of process 0, last bestmove process 1). Of course, it kind of implies that the bestmove wass send by process 0 before this, but maybe there is more output after that (e.g. a spurious start)?

**Comment by mooskagh** [comment]
Ok, here is equally fantastic scenario of being stuck with `searching=true`.
(as a bonus, the thread never starts the search, exactly as we saw!)

```
Uci Thread                     |  Worker thread        | searching
                               |                       |
                               |  In idle_loop():      |
                               |    searching = false  | false
                               |    wait(true)         |
                               |                       |
In start_searching:            |                       |
  searching = true             |   (system is too      | true
  notify_one()                 |    busy and cannot    |
                               |    wake up the        |
                               |    thread yet)        |
                               |                       |
In wait_for_search_finished(): |                       |
  wait(false)                  |
     wakes up from own         | 
     notify_one(), checks that |
     condition is false,       |
     falls asleep.             |
                               |   (thread can be
                               |    waked up but no
                               |    notification
                               |    anymore)                            
```

**Comment by vondele** [comment]
@mooskagh still trying to understand... but 2 (loosely related) comments:
- We have 3 kinds of threads, uci Thread (main process), mainThread, and worker threads. AFAICT, uciThread and mainThread communicate, mainThread wakes up workers.
- In principle, no uci commands can be send while threads search.

**Comment by noobpwnftw** [comment]
@mooskagh search.cpp line 262, the mainThread (UCI thread calls into) shouldn't wait on its own notify. During my attempt to implement a CV which tracks all waiting threads, it seems that each thread has its own instance of CV, `notify_all` won't change anything because they can only wait on their own thread.

**Comment by mooskagh** [comment]
Ok, so maybe I messed up the terminology.

So "Worker thread" in my last two comments is actually "mainThread", real worker threads don't play role in that memory dump at all, I'll rename it in the last diagram.

And UCI thread is waiting for "mainThread" to finish here: https://github.com/official-stockfish/Stockfish/blob/c9d73d1aa5cf609b626776a112cd699339fefb67/src/uci.cpp#L162

In the initial stack trace:
Thread #1 is UCI thread
Thread #2 is worker thread (seems irrelevant)
Thread #3 is mainThread

**Comment by mooskagh** [comment]
@noobpwnftw there is difference between "Thread object" and "system thread".
MainThread object is running in it's own system thread, but UCI system thread calls methods `start_searching()` and `wait_for_search_finished()` of MainThread object, which means that MainThread's cv is accessed from two system threads.

**Comment by snicolet** [comment]
In Thread::idle_loop we have this code
```C++
  while (true)
  {
      std::unique_lock<Mutex> lk(mutex);
      searching = false;
      cv.notify_one(); // Wake up anyone waiting for search finished
      cv.wait(lk, [&]{ return searching; });

      if (exit)
          return;

      lk.unlock();

      search();
  }
```

The lock is unlocked twice: once explicitly by the call `lk.unlock();`, once implicitly when lk goes out of scope at each turn of the loop. Correct?

**Comment by vondele** [comment]
@snicolet I don't think so, only once (goes out of scope with the return statement prior to unlock).

**Comment by vondele** [comment]
@mooskagh in the debug output, it would be a worker thread that doesn't start.

**Comment by Sopel97** [comment]
@snicolet  the unlock in the destructor is not called in this case https://en.cppreference.com/w/cpp/thread/unique_lock/%7Eunique_lock

**Comment by noobpwnftw** [comment]
@mooskagh I can see for any CV object, there can only have one outstanding wait operation from native threads. And the wake up itself isn't going to happen because:
thread A(UCI) wakes up thread B(mainThread) from `idle_loop` and in turn wakes up thread C(search) from `idle_loop`, both B and C gets `searching = true` set before receiving `notify_one`.
thread A then waits nothing, thread B after completing its search, wait for thread C to set its object's `searching = false`, then sets its own `searching = false` and wait for true. thread C after completing its search sets its own `searching = false` and wait for true.

From dump it shows that thread A(22400) failed to wake up thread B(6588) from `idle_loop`.

**Comment by noobpwnftw** [comment]
I can see one problem, if for all the time span between previous search end and `main()->start_searching();` firing another notify, thread B did not enter waiting state(`main()->wait_for_search_finished();` in the begin can pass without entering wait state which would require a notify from thread B), the new notify could get lost. However this is not the case because I'm using a Semaphore, so thread B must have been woken up, and the only place where `searching` is set to true is followed by a notify guarded in a mutex.
So it seems barriers don't really work that well when thread B gets notified, `searching` is still false and it entered another wait.

Does anyone have better reasons on how this should happen?

**Comment by vondele** [comment]
I'm travelling now but will re-join discussion in 2 days.

**Comment by mooskagh** [comment]
@vondele 
> @mooskagh in the debug output, it would be a worker thread that doesn't start.
mainThread is inherited from the Thread class and has the same start/stop mechanism, so I'd expect it to have similar issues.

In the particular memory dump it was mainThread which was hanging, that's why I name it so when talking about issue.

It may happen that non-main worker thread hang for different reason (unlikely), but as we also have mainThread hang, and that's the only memory dump that we have, I'd prefer to stick to it.

**Comment by mooskagh** [comment]
@noobpwnftw could you summarize where the problem was ever reproduced?
Is it both with mingw's STL and Visual Studio native STL?

If so, it looks extremely unlikely to me that it's a library bug.

**Comment by noobpwnftw** [comment]
I got reports saying using MSVC native STL can hang too, but I can't reproduce.
I can reliably get MINGW hangs using master, which is using a custom Mutex plus `std::conditional_variable_any`.
I can get very very rare hangs using MSVC with the same implementations above(by commenting the ifdef on _MSC_VER).
I can get more frequent hangs using MSVC with custom Mutex plus my custom CV implementation, which is what makes that dump.

I still fail to see how the Mutex and my CV implementation can fail on non-exotic cases(like it is not guarding dtors during waits).

**Comment by noobpwnftw** [comment]
From the dump and referring to the code, these are facts:
Only 2 threads involved, UCI and mainThread, the third one is a dummy thread created as uiThread.
1. UCI definitely is the only one that sets `searching = true` and within the same lock fired a notify.
2. mainThread doesn't wake up from `idle_loop` at all in the last iteration, by looking at variables like `nodes`, which only gets cleared at `start_thinking`, before 1.
3. That CV implementation guarantees `notify_one` always wakes up one waiter, if there is none, next wait will immediately success.
4. mainThread stuck at `idle_loop` while `searching = true`, which can only be explained by either the signal indeed got missed, or it didn't miss but the predicate failed.

**Comment by noobpwnftw** [comment]
@mooskagh You will probably like this one:

image
[Hang2.zip](https://github.com/official-stockfish/Stockfish/files/3619697/Hang2.zip)
https://github.com/noobpwnftw/Stockfish/tree/hang2

**Comment by mooskagh** [comment]
If you have mingw lying around, could you attach mingw.condition_variable.h file used in that version?

As you can see here
http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2007/n2406.html#gen_cond_var `conditional_variable_any` is really hard to get right..

That `.h` file contains multiple versions (for pre- and post- Windows XP systems). Do you know which of them is compiled?

Quick glance shows that there may be a problem, but I need to do more thorough look.
This is exceprt from https://github.com/meganz/mingw-std-threads/blob/master/mingw.condition_variable.h:
```cpp
class condition_variable_any
{
    condition_variable internal_cv_ {};
    windows7::mutex internal_mutex_ {};

    template<class L>
    bool wait_impl (L & lock, DWORD time)
    {
        unique_lock<decltype(internal_mutex_)> internal_lock(internal_mutex_);
        lock.unlock();
        bool success = internal_cv_.wait_impl(internal_lock, time);
        lock.lock();
        return success;
    }

    template<class L, class Predicate>
    void wait (L & lock, Predicate pred)
    {
        while (!pred())
            wait(lock, kInfinite);
    }

    void notify_one (void) noexcept
    {
        internal_cv_.notify_one();
    }
};
```

Cannot it be so:
Thread1:
```cpp
  while (true)
  {
      std::unique_lock<Mutex> lk(mutex);
      searching = false;
      cv.notify_one(); // Wake up anyone waiting for search finished
      cv.wait(lk, [&]{ return searching; });
      // Calls the following code:
       bool wait_impl (L & lock, DWORD time)
       {
           unique_lock<decltype(internal_mutex_)> internal_lock(internal_mutex_);
           lock.unlock();
           // <<<<---- Thread switch happens here!
           bool success = internal_cv_.wait_impl(internal_lock, time);
```
Switch to thread2, which notifies, but there's noone to notify.
```
void Thread::start_searching() {

  std::lock_guard<Mutex> lk(mutex);
  searching = true;
  cv.notify_one();
```


Even if `cv` has some sort of notification counter so that notification is not lost, at this point there's no protection against notifying own thread, it's just a counter. And if it happens (pretty realistic!) that Thread2 continutes executing until:
```cpp
void Thread::wait_for_search_finished() {

  std::unique_lock<Mutex> lk(mutex);
  cv.wait(lk, [&]{ return !searching; });
}
```
Then Thread2 just consumes all notifications.

**Comment by mooskagh** [comment]
I've just came to work, will only be able to take another look in ~9 hours.

**Comment by mooskagh** [comment]
In all cases that we reproduced, the search is unable to start. But in TCEC search started successfully but was not able to finish.
It looks like something different was happening?

**Comment by noobpwnftw** [comment]
[condition_variable.zip](https://github.com/official-stockfish/Stockfish/files/3619763/condition_variable.zip)
As `C:\msys64\mingw64\include\c++\9.2.0`.

**Comment by noobpwnftw** [comment]
@mooskagh same issue, it works both ways, just it is still hard to catch one with a dump.

**Comment by noobpwnftw** [comment]
I have experimented by adding a very long Sleep() in the exact point you suspected, but doesn't seem to break anything.

**Comment by mooskagh** [comment]
The file that you sent looks like a standard SGI STL file, and it's very different from mingw on that github page.. Could you check that that file is actually run? (i.e insert abort() and check whether it crashes)

**Comment by noobpwnftw** [comment]
@mooskagh In my entire MINGW installation, there are only 2 files that contains the implementation of `condition_variable_any`.
Another is here:
`C:\msys64\usr\lib\gcc\x86_64-pc-msys\9.1.0\include\c++\condition_variable`
[condition_variable.zip](https://github.com/official-stockfish/Stockfish/files/3619817/condition_variable.zip)

**Comment by noobpwnftw** [comment]
Now, looking back at the PR, in case of the notifier thread doesn't find a waiter, and a subsequent wait consumes the notify in the same thread, it will give one back, if the predicate success, no more notify should fire, if it fails, this will essentially turn the offending thread into a spin loop waiting on the other threads to enter wait, should this happen indefinitely, issue with OS scheduling(yielding added). Problem solved.

**Comment by noobpwnftw** [comment]
Now, now does this not bite us, like ever, until recently?

**Comment by Sopel97** [comment]
This looks like it can depend on behaviour of the internal cv.wait implementation, so it may be that it was just uncovered by a slightly different implementation.

**Comment by snicolet** [comment]
In my search_explosion2 branch I was referring to in my previous comment (https://github.com/official-stockfish/Stockfish/pull/2307#issuecomment-531842701 ) 

• we have the possibility to make each thread terminate at a different depth, so for instance we can have one worker thread terminate at depth 5, the main thread at depth 15 and the other workers at infinite depth
• there is a thread::thread_index() method which returns the number of the thread (0 is mainThread, >0 are worker threads), so we can add sleep() and yield() on a thread-by-thread basis 

==> Can these help us to simulate our scenarios in a deterministic manner?

**Comment by noobpwnftw** [comment]
To trigger what seems to be the case reliably, we have to do something so that:
A thread (UCI in my reproduced case) calls `start_searching`, B thread(MainThread) doesn't start to wait inside CV(unlocked external mutex, yet to wait on signal) before A thread enters `wait_for_search_finished`, which will either 1) `notify_one` in `start_searching` get lost due to no waiters, 2) A thread consumes all previous notifications. At this moment B thread enters wait and hang.

**Comment by noobpwnftw** [comment]
@snicolet I have made it reproducible: https://github.com/noobpwnftw/Stockfish/tree/hang2

**Comment by Tilps** [comment]
That is a pretty bad condition_variable implementation in your reproducible hang :)

**Comment by noobpwnftw** [comment]
To make the hang "disappear", move the `std::unique_lock<Mutex> lk(lock);` before unlocking external lock probably will do, however, this is not required for CV implementations, AFAIK.

@Tilps of course, as long as it serves its purpose.

**Comment by Tilps** [comment]
condition variable implementation requires that unlocking and adding to the waiter queue is atomic - without the lock being moved before the unlock that definitely isn't true.

**Comment by noobpwnftw** [comment]
Not with this implementation apparently:
https://sourceforge.net/p/mingw-w64/mingw-w64/ci/b2302cd60b94a6d68a0e18b3f886b82f4b7f22b7/tree/mingw-w64-libraries/winpthreads/src/cond.c

And they seems to have it fixed recently:
https://sourceforge.net/p/mingw-w64/mingw-w64/ci/330025c54b85512d54b6960fad07498365c8fee3/

**Comment by Tilps** [comment]
I don't have extreme confidence, but yes the fix looks useful.

**Comment by noobpwnftw** [comment]
They implement the waiting object as a global semaphore, in this particular version of `libwinpthread`, there is no lock guard on OS level to ensure atomic releasing and waiting, although there is a `waiters_count_lock_` but that basically does nothing. The actual fix is this: https://sourceforge.net/p/mingw-w64/mingw-w64/ci/1ebdd63cfffd936c97443f58216a0a09a9e721ee/
which binds the behavior with the external lock.

**Comment by snicolet** [comment]
My opinion is that if some condition variable implementations are buggy and don't respect the specification, then we should not use condition variables at all, and not try to play the trick game of tweaking the predicate condition to turn around their bugs in external librairies.

There must be another simple way to solve the thread synchronization problem in Stockfish without using condition variables.

**Comment by noobpwnftw** [comment]
So I guess there are still plenty of people who knows nothing about how to implement locks like myself.

The thing is, I can take all time flaming them for the bug in the wild(this is going to take a while for everyone to upgrade), but we can still adapt with trivial changes.

IMO, there are two ways:
1) It doesn't break anything by firing extra `notify_one`s regardless of bug.
2) Custom CV implementation, since we use custom Mutex already.

I'd prefer 1), for 2) if next time if it is the `steady_clock` that becomes unreliable, and we will remove all timers or implement our own? (I consider the removal of conditional variables equivalent to implement our own CV).

**Comment by mooskagh** [comment]
`notify_one()` may mask bug and make it much more rare, but not fix.
If there is a bug, I'd prefer to make it more frequent, not more rare, so that it's easier to catch and fix properly.

As for custom CV implementation, it's pretty hard to get right, see http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2007/n2406.html#gen_cond_var

So I'm against either.

By the way, why aren't we using std::mutex (and std::condition_variable rather than std::condition_variable_any)?




Also, with which STL was this https://github.com/official-stockfish/Stockfish/pull/2307#issuecomment-531904809 compiled?
If it's with visual studio STL, and it also hangs there (i.e. we have a hang in two completely different STL implementations), why are we searching for bugs in mingw?

**Comment by noobpwnftw** [comment]
Custom mutexes are only used on MINGW, because theirs sucks.
That was with my semaphore CV, probably has the same bug as MINGW.
I will try to reproduce a hang with only Mutex + `std::conditional_variable_any` on MSVC, if there is any.

**Comment by noobpwnftw** [comment]
It looks like there is no hang with MSVC's CV, only MINGW and my custom ones.

**Comment by mooskagh** [comment]
How exactly do mingw mutices suck? Are they just slow? Is it a problem, aren't they only locked once per search session per thread?

**Comment by noobpwnftw** [comment]
Also used in synchronizing `cout`.

**Comment by noobpwnftw** [comment]
So, no problems at least.
http://tests.stockfishchess.org/tests/view/5d80958f0ebc5971531d0a1b
http://tests.stockfishchess.org/tests/view/5d8095c60ebc5971531d0a1d

**Comment by noobpwnftw** [comment]
Inside MINGW's runtime, `pthread_create` uses CV, some allocators too, there are more.
I think removing CV is not a more "proper" tweak than this one, unless you'd implement everything else too.

Third option, ditch MINGW support entirely, tell them to switch to WSL or something.

**Comment by noobpwnftw** [comment]
Closing this, since this will not fix everything else inside bugged MINGW.

**Comment by snicolet** [comment]
Just for reference, here is a version using atomic<bool> instead of condition variable: https://github.com/snicolet/Stockfish/tree/hang

Does it hang?

**Comment by noobpwnftw** [comment]
Do you want to test it under MINGW?

**Comment by snicolet** [comment]
> It does not hang for me, MINGW.

Thanks!
I am also interested in your hang2 branch, especially because you said you had a deterministic hang with your custom condition variable implementation with sloppy notify_one.

However, you deleted the hang2 branch, so I am no longer able to try to look at the scenario for the repeatable hang. I would like to try to simulate it with the atomic<bool> branch I have.

Would it be possible for you to undelete the branch, and republish it? :-) 
Recipes for how to do that with Git: https://stackoverflow.com/questions/3640764/can-i-recover-a-branch-after-its-deletion-in-git , it it a combinaison of **git reflog** and **git checkout -b**

**Comment by noobpwnftw** [comment]
That is gone, I made a new branch, in there:
A reconstructed code that mimics MINGW's way of doing things. Hangs 100% with bench.
An early CV implementation of mine that fails the same way, and how to make it right.
https://github.com/noobpwnftw/Stockfish/tree/hang3

**Comment by snicolet** [comment]
Superb, thanks, I will have a look!

**Comment by snicolet** [comment]
I have try to get a reduction of the hang there: https://github.com/official-stockfish/Stockfish/issues/2309
